### ISSUE TITLE
Slovakian, Czech, Russian, Polish special characters missing...

### ISSUE THREAD

**Issue by Spikeone** [Return-To-The-Roots/s25client#79]
1. The slovakian language needs following characters:
   ĺ ľ š č ť ž ý á í é ä ô ď ŕ ř ě ň

Following do not work:
ĺ ľ š č ť ž ä ô ď ŕ ř ě ň
1. The polish language need following characters:

Ą - U+0104
ą - U+0105
Ć - U+0106
ć - U+0107
Ę - U+0118
ę - U+0119
Ł - U+0141
ł - U+0142
Ń - U+0143
ń - U+0144
Ó - U+00D3
ó - U+00F3 (that works ok)
Ś - U+015A
ś - U+015B
Ź - U+0179
ź - U+017A
Ż - U+017B
ż - U+017C

https://bugs.launchpad.net/s25rttr/+bug/371755

Edit by @Flamefire: Removed outdated discussion.

Current status: We are using UTF8 fonts ingame so every character is possible. We need a graphic for each name as U+xxxx.bmp in S2 palette colors and include them into the FONTS.LST. See https://github.com/Return-To-The-Roots/s25client/wiki/How-to-edit-fonts

For testing you can simply type the chars into any edit control (e.g. lobby login)

**Comment by Flamefire** [comment]
Can anyone summarize and clean this up? Best @Spikeone so he can add what he still has. 
I think this should go into the next stable as missing translations for a couple of languages (and especially as it is not possible to easily fix this as it is right now) is quite a big thing. Also makes a nice announcement

**Comment by Flamefire** [comment]
@Flow86: Should we use UTF-8 or Unicode as the filename for the letters? Unicode seems to be the best option: Hexadecimal with 4 digits: 0010, 00AF, 01AF, ...

**Comment by Spikeone** [comment]
When I first started, flosoft wanted unicode

**Comment by Flow86** [comment]
yes uncode

**Comment by Flamefire** [comment]
Support for more glyphs is added. Glyphs are in the works.

**Comment by Spikeone** [comment]
After such a long time of promising I finally did something.

I started with the slovakian letters, used this site as reference: http://www.geocities.ws/click2speak/unicode/chars_sk.html

I created all capital letters first but now I need someone who tests this as I'm not sure if everything is correct (also regarding positions).

http://www.rfmonline.de/rttr/spike/outline_fonts.rar

**Comment by tehKaiN** [comment]
Hi there,

I've been working on PL fonts around 2015, but It looks like I've never commited them. So, here they are:
[http://piwnica.ws/_stuff/rttr-pl-font.zip](http://piwnica.ws/_stuff/rttr-pl-font.zip)
They are ripped from S2 ofc and are in yellow palette, but it should be easy enough for someone to port it to proper palette.
Here is reference img from S2: http://i.imgur.com/JhcFkQp.png

Good to see this project is still alive, I'll try to contribute from time to time :)

**Comment by Flamefire** [comment]
Any progress here so we can add the new chars?
@tehKaiN Can you rename the files to their unicode codepoints? They should be `U+xxxx.bmp` Also png files won't work as the compression probably messed up the colors. For more information on editing fonts see https://github.com/Return-To-The-Roots/s25client/wiki/How-to-edit-fonts

**Comment by frin** [comment]
I have recently added Slovenian characters, that is č, š, ž (Č, Š, Ž uppercase) which have been merged, so you can scratch those off the list. Some of these are also used in other Slavic languages.

A little advice to others editing these fonts: MS Paint messes up the colors when drawing pixels (at least in Windows 7), using Photoshop does work great, just merge the layers before saving.

Here is a useful UTF8 table: http://www.utf8-chartable.de/unicode-utf8-table.pl?start=256&unicodeinhtml=dec

**Comment by Flamefire** [comment]
@Spikeone Could you make a list with missing characters (updated) and their unicode codepoints like above for polish?

If you can even create some of them, it would be great. Otherwise we could add replacements from the existing ones in the meantime (like `l` for `ľ`)

**Comment by Spikeone** [comment]
Which font sizes does rttr use? 0,1 and 2 or only 2?

**Comment by Flamefire** [comment]
All 3.

**Comment by izikeros** [comment]
I have created `fonts.lst` where added files for polish characters. I have copied characters without accents, "ogonki", etc... as placeholder for missing glyphs (e.g. "e" serves as "ę", "L" as "Ł"...). It is a temporary solution but looks much better than missing characters. It covers all three font sizes. For some (short) time it should be available here: https://file.io/bSRRXfSif0Gq. Please use it if that is fine.

**Comment by Flow86** [comment]
[fonts.zip](https://github.com/Return-To-The-Roots/s25client/files/5835794/fonts.zip)

added it as a zip here

**Comment by Flow86** [comment]
@Flamefire  that would be an option to add some code to copy the glyphs on load? if they are missing? that would probably be an interesting way for now?

**Comment by Flamefire** [comment]
We are already using a semi-custom format for the (extended) fonts and given the range of possible unicode characters (about 65k at least, actually over 100k) it would certainly make sense to add some more functionality and e.g. save space.
However there are limits. E.g. support for combining characters (stuff like ^ + a = â) will be nearly impossible to implement with the current system.
Replacement characters would make sense too, either by deduplicating images similar to the `offsets` in the text archives or by an attached mapping format similar to the palettes. For the former I'd use an invalid "bobtype" in the font archive which means that an uint32 is following containing the char that should be duplicated and copy the image (wastes some space obviously, but well...) Or we append the mapping and handle it in the char-to-positionInTexture mapping later only.

In any case someone needs to do that mapping. And easiest is likely copying files and de-duplicating on writing

**Comment by lstolcman** [comment]
Any progress here (or what is needed to commit at least small part, e.g. polish glyphs - the part I'm interested in)?
The 0.9.1 - version released 8 years after the previous one - still has this "bug" 😕 

![obraz](https://user-images.githubusercontent.com/4583553/147393610-c7eb7ee7-834c-4b85-85ae-d32b7f491229.png)

**Comment by Spikeone** [comment]
@lstolcman to add "glyphs" you can refer to this article: https://github.com/Return-To-The-Roots/s25client/wiki/How-to-edit-fonts

**Comment by tehKaiN** [comment]
Hi,

someone have recently (january this year) contacted me about this because my url with glyphs is dead and now I see the issue is still alive, so I'll post them here:

[rttr-pl-font.zip](https://github.com/Return-To-The-Roots/s25client/files/7776806/rttr-pl-font.zip)

Also, there was something wrong with those fonts size- or color-wise and they needed adjustment, but I gave up at this point because I'm no graphician and wouldn't dare to commit messed up fonts.

Here's the DOSBox screenshot of the official Polish Settlers 2 Gold release with readme changed so that it neatly contains all Polish characters. Maybe it will be useful for someone here. As one can see some fonts (particularly uppercase Ó, Ź, Ś, Ń are not that good).

![all](https://user-images.githubusercontent.com/16180344/147404581-4bd6c01e-ab3e-40b7-b321-f618fda20cae.png)

Since then I've tried to put this issue into attention of my fellow pixel artists, but sadly so far no luck. :(
